build_config/lib/src/build_config.dart

```diff
@@ -0,0 +1,344 @@
+// Copyright (c) 2017, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:async';
+import 'dart:io';
+
+import 'package:path/path.dart' as p;
+import 'package:yaml/yaml.dart';
+
+import 'pubspec.dart';
+
+/// The parsed values from a `build.yaml` file.
+class BuildConfig {
+  /// Supported values for the `platforms` attribute.
+  static const _allPlatforms = const [_vmPlatform, _webPlatform];
+  static const _vmPlatform = 'vm';
+  static const _webPlatform = 'web';
+
+  static const _targetOptions = const [
+    _builders,
+    _default,
+    _dependencies,
+    _excludeSources,
+    _generateFor,
+    _platforms,
+    _sources,
+  ];
+  static const _builders = 'builders';
+  static const _default = 'default';
+  static const _dependencies = 'dependencies';
+  static const _excludeSources = 'exclude_sources';
+  static const _generateFor = 'generate_for';
+  static const _platforms = 'platforms';
+  static const _sources = 'sources';
+
+  static const _builderOptions = const [
+    _builderFactories,
+    _import,
+    _inputExtension,
+    _outputExtensions,
+    _target,
+  ];
+  static const _builderFactories = 'builder_factories';
+  static const _import = 'import';
+  static const _inputExtension = 'input_extension';
+  static const _outputExtensions = 'output_extensions';
+  static const _target = 'target';
+
+  /// Returns a parsed [BuildConfig] file in [path], if one exists.
+  ///
+  /// Otherwise uses the default setup.
+  static Future<BuildConfig> fromPackageDir(Pubspec pubspec, String path,
+      {bool includeWebSources: false}) async {
+    final configPath = p.join(path, 'build.yaml');
+    final file = new File(configPath);
+    if (await file.exists()) {
+      return new BuildConfig.parse(pubspec, await file.readAsString(),
+          includeWebSources: includeWebSources);
+    } else {
+      return new BuildConfig.useDefault(pubspec,
+          includeWebSources: includeWebSources);
+    }
+  }
+
+  final String packageName;
+
+  /// All the `builders` defined in a `build.yaml` file.
+  final dartBuilderBinaries = <String, BuilderDefinition>{};
+
+  /// All the `targets` defined in a `build.yaml` file.
+  final dartLibraries = <String, BuildTarget>{};
```

Comment:
I wonder if we actually still want this?

Reply:
We probably don't need it for `build_runner` but I think we still need it for `dazel` and my plan was to eventually rip out the implementation from there and wire this one in.

I should rename these though...